agents/vmagent/vmagent.go

```diff
@@ -0,0 +1,183 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vmagent
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"runtime/pprof"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/percona/pmm-agent/agents/process"
+	"github.com/percona/pmm-agent/config"
+)
+
+// VMAgent represents VictoriaMetrics agent.
+type VMAgent struct {
+	remoteInsecure   bool
+	remoteUserName   string
+	remotePassword   string
+	binaryPath       string
+	client           *http.Client
+	remoteWriteURL   *url.URL
+	listenURL        *url.URL
+	scrapeConfigPath string
+	lastCfg          []byte
+	tmpDir           string
+	l                *logrus.Entry
+	cancel           context.CancelFunc
+	done             chan struct{}
+}
+
+// NewVMAgent - creates vmagent object and scrape config file.
+func NewVMAgent(cfg *config.Config) (*VMAgent, error) {
+	remoteWriteURL := *cfg.Server.URL()
+	remoteWriteURL.Path = path.Join(remoteWriteURL.Path, "victoriametrics", "api", "v1", "write")
+	remoteWriteURL.User = nil
+	scrapeCfgPath := path.Join(cfg.Paths.TempDir, "vmagent-scrape-config.yaml")
+	scrapeCfg, err := ioutil.ReadFile(scrapeCfgPath) //nolint:gosec
```

Comment:
Potentially, it can be useful during restarts:
1) read content of config file (add it to cached data)
2) start vmagent
3) get cfg from pmm-server
4) if cachedcfg == newCfg, do nothing.

 It allows to continue targets scraping w/o pmm-server connection.

Reply:
But it can led us to pass some metrics related to exporter1 as metrics related to exporter2, because after restart they may have different ports.

Reply:
Added tempFile